src/_pytest/_code/code.py

```diff
@@ -572,8 +572,13 @@ def match(self, regexp):
         raised.
         """
         __tracebackhide__ = True
-        if not re.search(regexp, str(self.value)):
-            assert 0, "Pattern '{!s}' not found in '{!s}'".format(regexp, self.value)
+        value = (
+            text_type(self.value) if isinstance(regexp, text_type) else str(self.value)
+        )
+        if not re.search(regexp, value):
+            raise AssertionError(
+                u"Pattern '{}' not found in '{}'".format(regexp, value)
```

Comment:
In python2 this will raise on non-ascii bytes

Reply:
Hmm indeed. Any suggestions?

I guess we can use `safe_str` to display the messages? Although `safe_str` garbles on invalid output, which might end up being more confusing.

Alternatively we can try to decode it, and upon failure, just show `<decode error>` instead of the original string:

```
Pattern 'hello' not found in <<decode error>>
```

Or something like this.

Reply:
!r might be the best bet :S

Reply:
Done using `{!r}` for formatting instead, which is the safest approach I think.

Reply:
> Done using `{!r}` for formatting instead, which is the safest approach I think.

I think that change will need to be ported to pytest 5

Reply:
probably should, yeah -- there's currently this weird output due to not using `!r`

```
E           AssertionError: Pattern ''foo' not found in ''bar'
```

```python
import pytest


def test():
    with pytest.raises(AssertionError, match="'foo"):
        raise AssertionError("'bar")
```

Reply:
I don't think this should be ported to pytest 5... this is a non-issue on Python 3, I believe.

Reply:
I created #5553